### ISSUE TITLE
hardcoded `WriteTimeout` cause unexpected error in a slow flow

### ISSUE THREAD

**Issue by CNLHC** [ory/kratos#2926]
### Preflight checklist

- [X] I could not find a solution in the existing issues, docs, nor discussions.
- [X] I agree to follow this project's [Code of Conduct](https://github.com/ory/kratos/blob/master/CODE_OF_CONDUCT.md).
- [X] I have read and am following this repository's [Contribution Guidelines](https://github.com/ory/kratos/blob/master/CONTRIBUTING.md).
- [ ] This issue affects my [Ory Network](https://www.ory.sh/) project.
- [ ] I have joined the [Ory Community Slack](https://slack.ory.sh).
- [ ] I am signed up to the [Ory Security Patch Newsletter](https://ory.us10.list-manage.com/subscribe?u=ffb1a878e4ec6c0ed312a3480&id=f605a41b53).

### Describe the bug

> TLDR: Sometimes the OIDC callback may be invoked twice by the browser and cause unexpected errors.

Edit: This problem is caused by the hardcoded `WriteTimeout` parameters in https://github.com/ory/graceful/blob/927f8152c87148e97688d02bc0abd44d75f4954a/http_defaults.go#L72.

When the user invokes an OIDC flow, the provider will redirect to the `/self-service/methods/oidc/callback/:provier`
which will invoke https://github.com/ory/kratos/blob/8406eaf92006d9812108bd3ae57245f01e627bfc/selfservice/strategy/oidc/strategy.go#L306

In some situations, the OIDC provider may take a while to perform the token exchange. During token exchanging, the browser receives nothing (even the HTTP header) and maybe resends the same request to the server again and causes an error.

I do not find the document about this retry behavior.  According to my test, if a browser sends a request and receives nothing in 10 seconds, it will try to resend the request. 

### Reproducing the bug

1. prepare a slow OIDC provider ( I add a `time.Sleep(10 * time. Second)` to an OIDC provider code for the test)
2. create a login or registration flow
3. the browser will be redirected to the error page, the error info will be like

```
Resumable ID from cookie could not be found in the datastore
```

### Relevant log output

Here is the full relevant log output.

```shell
INFO[2022-12-05T17:21:41+08:00]<MY_PATH>go/pkg/mod/github.com/ory/x@v0.0.513/reqlog/middleware.go:134 github.com/ory/x/reqlog.(*Middleware).ServeHTTP() started handling request                      http_request=map[headers:map[<COMMON_HEADER> cookie:[<CSRF_TOKEN_KEY><CSRF_TOKEN_VAL>=; ory_kratos_continuity=<COOKIE_TOKEN1>] dnt:1 sec-ch-ua:"<UA>] host:localhost:4433 method:GET path:/self-service/methods/oidc/callback/gitlab query:code=<MY_CODE>&state=<STATE_TOKEN1> remote:[::1]:60592 scheme:http]
DEBU[2022-12-05T17:22:02+08:00]<MY_PATH>Project/FOSS/kratos/selfservice/flow/login/hook.go:136 github.com/ory/kratos/selfservice/flow/login.(*HookExecutor).PostLoginHook() Running ExecuteLoginPostHook.                 audience=application flow_method=oidc http_request=map[headers:map[<COMMON_HEADER> cookie:[<CSRF_TOKEN_KEY><CSRF_TOKEN_VAL>=; ory_kratos_continuity=<COOKIE_TOKEN1>] dnt:1 sec-ch-ua:"<UA>] host:localhost:4433 method:GET path:/self-service/methods/oidc/callback/gitlab query:code=<MY_CODE>&state=<STATE_TOKEN1> remote:[::1]:60592 scheme:http] identity_id=<IDENTITY_UUID>service_name=Ory Kratos service_version=master
INFO[2022-12-05T17:22:02+08:00]<MY_PATH>Project/FOSS/kratos/selfservice/flow/login/hook.go:199 github.com/ory/kratos/selfservice/flow/login.(*HookExecutor).PostLoginHook() Identity authenticated successfully and was issued an Ory Kratos Session Cookie.  audience=audit http_request=map[headers:map[<COMMON_HEADER> cookie:[<CSRF_TOKEN_KEY><CSRF_TOKEN_VAL>=; ory_kratos_continuity=<COOKIE_TOKEN1>] dnt:1 sec-ch-ua:"<UA>] host:localhost:4433 method:GET path:/self-service/methods/oidc/callback/gitlab query:code=<MY_CODE>&state=<STATE_TOKEN1> remote:[::1]:60592 scheme:http] identity_id=<IDENTITY_UUID>service_name=Ory Kratos service_version=master session_id=90dc01f4-cf2b-42ac-a16d-c5ffcb35bf99
INFO[2022-12-05T17:22:02+08:00]<MY_PATH>go/pkg/mod/github.com/ory/x@v0.0.513/reqlog/middleware.go:146 github.com/ory/x/reqlog.(*Middleware).ServeHTTP() completed handling request                    http_request=map[headers:map[<COMMON_HEADER> cookie:[<CSRF_TOKEN_KEY><CSRF_TOKEN_VAL>=; ory_kratos_continuity=<COOKIE_TOKEN1>] dnt:1 sec-ch-ua:"<UA>] host:localhost:4433 method:GET path:/self-service/methods/oidc/callback/gitlab query:code=<MY_CODE>&state=<STATE_TOKEN1> remote:[::1]:60592 scheme:http] http_response=map[headers:map[<CACHE_HEADER> location:http://localhost:3000/ set-cookie:[ory_kratos_continuity=MTY3MDIzMjEwMXxEdi1CQkFFQ180SUFBUkFCRUFBQUJQLUNBQUE9fI1rWkD8bjafhmYvXPvhw2BkfK3pB5eC0hFLybCtbfLc; Path=/; Expires=Wed, 04 Jan 2023 09:21:41 GMT; Max-Age=2592000; HttpOnly; SameSite=Lax <CSRF_TOKEN_KEY>=LAAd83iH5n2fAJcEVEjGUmkZOy9HR0xQmo1L5HGiTKM=; Path=/; Max-Age=31536000; HttpOnly; SameSite=Lax ory_kratos_session=<SESSION_COOKIE>; Path=/; Expires=Tue, 06 Dec 2022 09:22:01 GMT; Max-Age=86399; HttpOnly; SameSite=Lax] vary:Cookie] size:49 status:303 text_status:See Other took:21.510034625s]
INFO[2022-12-05T17:22:02+08:00]<MY_PATH>go/pkg/mod/github.com/ory/x@v0.0.513/reqlog/middleware.go:134 github.com/ory/x/reqlog.(*Middleware).ServeHTTP() started handling request                      http_request=map[headers:map[<COMMON_HEADER> cookie:[<CSRF_TOKEN_KEY><CSRF_TOKEN_VAL>=; ory_kratos_continuity=<COOKIE_TOKEN1>] dnt:1 sec-ch-ua:"<UA>] host:localhost:4433 method:GET path:/self-service/methods/oidc/callback/gitlab query:code=<MY_CODE>&state=<STATE_TOKEN1> remote:[::1]:60700 scheme:http]
ERRO[2022-12-05T17:22:02+08:00]<MY_PATH>go/pkg/mod/github.com/ory/x@v0.0.513/logrusx/helper.go:125 github.com/ory/x/logrusx.(*Logger).Logf() An error occurred and is being forwarded to the error user interface.  audience=application error=map[debug:Resumable ID from cookie could not be found in the datastore: id=
rid=
error=Unable to locate the resource
reason=
details=map[]
debug=

github.com/ory/x/sqlcon.HandleError
        <MY_PATH>go/pkg/mod/github.com/ory/x@v0.0.513/sqlcon/error.go:76
github.com/ory/kratos/persistence/sql.(*Persister).GetContinuitySession
        <MY_PATH>Project/FOSS/kratos/persistence/sql/persister_continuity.go:36
github.com/ory/kratos/continuity.(*ManagerCookie).container
        <MY_PATH>Project/FOSS/kratos/continuity/manager_cookie.go:117
github.com/ory/kratos/continuity.(*ManagerCookie).Continue
        <MY_PATH>Project/FOSS/kratos/continuity/manager_cookie.go:67
github.com/ory/kratos/selfservice/strategy/oidc.(*Strategy).validateCallback
        <MY_PATH>Project/FOSS/kratos/selfservice/strategy/oidc/strategy.go:268
github.com/ory/kratos/selfservice/strategy/oidc.(*Strategy).handleCallback
        <MY_PATH>Project/FOSS/kratos/selfservice/strategy/oidc/strategy.go:312
github.com/ory/kratos/selfservice/strategy.disabledWriter
        <MY_PATH>Project/FOSS/kratos/selfservice/strategy/handler.go:28
github.com/ory/kratos/selfservice/strategy.IsDisabled.func1
        <MY_PATH>Project/FOSS/kratos/selfservice/strategy/handler.go:33
github.com/ory/kratos/x.NoCacheHandle.func1
        <MY_PATH>Project/FOSS/kratos/x/nocache.go:21
github.com/ory/kratos/x.NoCacheHandle.func1
        <MY_PATH>Project/FOSS/kratos/x/nocache.go:21
github.com/julienschmidt/httprouter.(*Router).ServeHTTP
        <MY_PATH>go/pkg/mod/github.com/julienschmidt/httprouter@v1.3.0/router.go:387
github.com/ory/nosurf.(*CSRFHandler).handleSuccess
        <MY_PATH>go/pkg/mod/github.com/ory/nosurf@v1.2.7/handler.go:234
github.com/ory/nosurf.(*CSRFHandler).ServeHTTP
        <MY_PATH>go/pkg/mod/github.com/ory/nosurf@v1.2.7/handler.go:191
github.com/urfave/negroni.Wrap.func1
        <MY_PATH>go/pkg/mod/github.com/urfave/negroni@v1.0.0/negroni.go:46
github.com/urfave/negroni.HandlerFunc.ServeHTTP
        <MY_PATH>go/pkg/mod/github.com/urfave/negroni@v1.0.0/negroni.go:29
github.com/urfave/negroni.middleware.ServeHTTP
        <MY_PATH>go/pkg/mod/github.com/urfave/negroni@v1.0.0/negroni.go:38
github.com/ory/kratos/x.glob..func1
        <MY_PATH>Project/FOSS/kratos/x/clean_url.go:15
github.com/urfave/negroni.HandlerFunc.ServeHTTP
        <MY_PATH>go/pkg/mod/github.com/urfave/negroni@v1.0.0/negroni.go:29
github.com/urfave/negroni.middleware.ServeHTTP
        <MY_PATH>go/pkg/mod/github.com/urfave/negroni@v1.0.0/negroni.go:38
net/http.HandlerFunc.ServeHTTP
        /opt/homebrew/Cellar/go/1.19/libexec/src/net/http/server.go:2109
github.com/prometheus/client_golang/prometheus/promhttp.InstrumentHandlerResponseSize.func1
        <MY_PATH>go/pkg/mod/github.com/prometheus/client_golang@v1.13.0/prometheus/promhttp/instrument_server.go:284
net/http.HandlerFunc.ServeHTTP
        /opt/homebrew/Cellar/go/1.19/libexec/src/net/http/server.go:2109
github.com/prometheus/client_golang/prometheus/promhttp.InstrumentHandlerCounter.func1
        <MY_PATH>go/pkg/mod/github.com/prometheus/client_golang@v1.13.0/prometheus/promhttp/instrument_server.go:142
net/http.HandlerFunc.ServeHTTP
        /opt/homebrew/Cellar/go/1.19/libexec/src/net/http/server.go:2109
github.com/prometheus/client_golang/prometheus/promhttp.InstrumentHandlerDuration.func1
        <MY_PATH>go/pkg/mod/github.com/prometheus/client_golang@v1.13.0/prometheus/promhttp/instrument_server.go:92
net/http.HandlerFunc.ServeHTTP
        /opt/homebrew/Cellar/go/1.19/libexec/src/net/http/server.go:2109
github.com/prometheus/client_golang/prometheus/promhttp.InstrumentHandlerDuration.func2
        <MY_PATH>go/pkg/mod/github.com/prometheus/client_golang@v1.13.0/prometheus/promhttp/instrument_server.go:104
net/http.HandlerFunc.ServeHTTP
        /opt/homebrew/Cellar/go/1.19/libexec/src/net/http/server.go:2109
github.com/prometheus/client_golang/prometheus/promhttp.InstrumentHandlerRequestSize.func1
        <MY_PATH>go/pkg/mod/github.com/prometheus/client_golang@v1.13.0/prometheus/promhttp/instrument_server.go:234
net/http.HandlerFunc.ServeHTTP
        /opt/homebrew/Cellar/go/1.19/libexec/src/net/http/server.go:2109
github.com/ory/x/prometheusx.Metrics.instrumentHandlerStatusBucket.func1
        <MY_PATH>go/pkg/mod/github.com/ory/x@v0.0.513/prometheusx/metrics.go:111
net/http.HandlerFunc.ServeHTTP
        /opt/homebrew/Cellar/go/1.19/libexec/src/net/http/server.go:2109 message:no resumable session found reason:The browser does not contain the necessary cookie to resume the session. This is a security violation and was blocked. Please clear your browser's cookies and cache and try again! stack_trace:
github.com/ory/kratos/continuity.(*ManagerCookie).container
        <MY_PATH>Project/FOSS/kratos/continuity/manager_cookie.go:124
github.com/ory/kratos/continuity.(*ManagerCookie).Continue
        <MY_PATH>Project/FOSS/kratos/continuity/manager_cookie.go:67
github.com/ory/kratos/selfservice/strategy/oidc.(*Strategy).validateCallback
        <MY_PATH>Project/FOSS/kratos/selfservice/strategy/oidc/strategy.go:268
github.com/ory/kratos/selfservice/strategy/oidc.(*Strategy).handleCallback
        <MY_PATH>Project/FOSS/kratos/selfservice/strategy/oidc/strategy.go:312
github.com/ory/kratos/selfservice/strategy.disabledWriter
        <MY_PATH>Project/FOSS/kratos/selfservice/strategy/handler.go:28
github.com/ory/kratos/selfservice/strategy.IsDisabled.func1
        <MY_PATH>Project/FOSS/kratos/selfservice/strategy/handler.go:33
github.com/ory/kratos/x.NoCacheHandle.func1
        <MY_PATH>Project/FOSS/kratos/x/nocache.go:21
github.com/ory/kratos/x.NoCacheHandle.func1
        <MY_PATH>Project/FOSS/kratos/x/nocache.go:21
github.com/julienschmidt/httprouter.(*Router).ServeHTTP
        <MY_PATH>go/pkg/mod/github.com/julienschmidt/httprouter@v1.3.0/router.go:387
github.com/ory/nosurf.(*CSRFHandler).handleSuccess
        <MY_PATH>go/pkg/mod/github.com/ory/nosurf@v1.2.7/handler.go:234
github.com/ory/nosurf.(*CSRFHandler).ServeHTTP
        <MY_PATH>go/pkg/mod/github.com/ory/nosurf@v1.2.7/handler.go:191
github.com/urfave/negroni.Wrap.func1
        <MY_PATH>go/pkg/mod/github.com/urfave/negroni@v1.0.0/negroni.go:46
github.com/urfave/negroni.HandlerFunc.ServeHTTP
        <MY_PATH>go/pkg/mod/github.com/urfave/negroni@v1.0.0/negroni.go:29
github.com/urfave/negroni.middleware.ServeHTTP
        <MY_PATH>go/pkg/mod/github.com/urfave/negroni@v1.0.0/negroni.go:38
github.com/ory/kratos/x.glob..func1
        <MY_PATH>Project/FOSS/kratos/x/clean_url.go:15
github.com/urfave/negroni.HandlerFunc.ServeHTTP
        <MY_PATH>go/pkg/mod/github.com/urfave/negroni@v1.0.0/negroni.go:29
github.com/urfave/negroni.middleware.ServeHTTP
        <MY_PATH>go/pkg/mod/github.com/urfave/negroni@v1.0.0/negroni.go:38
net/http.HandlerFunc.ServeHTTP
        /opt/homebrew/Cellar/go/1.19/libexec/src/net/http/server.go:2109
github.com/prometheus/client_golang/prometheus/promhttp.InstrumentHandlerResponseSize.func1
        <MY_PATH>go/pkg/mod/github.com/prometheus/client_golang@v1.13.0/prometheus/promhttp/instrument_server.go:284
net/http.HandlerFunc.ServeHTTP
        /opt/homebrew/Cellar/go/1.19/libexec/src/net/http/server.go:2109
github.com/prometheus/client_golang/prometheus/promhttp.InstrumentHandlerCounter.func1
        <MY_PATH>go/pkg/mod/github.com/prometheus/client_golang@v1.13.0/prometheus/promhttp/instrument_server.go:142
net/http.HandlerFunc.ServeHTTP
        /opt/homebrew/Cellar/go/1.19/libexec/src/net/http/server.go:2109
github.com/prometheus/client_golang/prometheus/promhttp.InstrumentHandlerDuration.func1
        <MY_PATH>go/pkg/mod/github.com/prometheus/client_golang@v1.13.0/prometheus/promhttp/instrument_server.go:92
net/http.HandlerFunc.ServeHTTP
        /opt/homebrew/Cellar/go/1.19/libexec/src/net/http/server.go:2109
github.com/prometheus/client_golang/prometheus/promhttp.InstrumentHandlerDuration.func2
        <MY_PATH>go/pkg/mod/github.com/prometheus/client_golang@v1.13.0/prometheus/promhttp/instrument_server.go:104
net/http.HandlerFunc.ServeHTTP
        /opt/homebrew/Cellar/go/1.19/libexec/src/net/http/server.go:2109
github.com/prometheus/client_golang/prometheus/promhttp.InstrumentHandlerRequestSize.func1
        <MY_PATH>go/pkg/mod/github.com/prometheus/client_golang@v1.13.0/prometheus/promhttp/instrument_server.go:234
net/http.HandlerFunc.ServeHTTP
        /opt/homebrew/Cellar/go/1.19/libexec/src/net/http/server.go:2109
github.com/ory/x/prometheusx.Metrics.instrumentHandlerStatusBucket.func1
        <MY_PATH>go/pkg/mod/github.com/ory/x@v0.0.513/prometheusx/metrics.go:111
net/http.HandlerFunc.ServeHTTP
        /opt/homebrew/Cellar/go/1.19/libexec/src/net/http/server.go:2109
github.com/ory/x/prometheusx.(*MetricsManager).ServeHTTP
        <MY_PATH>go/pkg/mod/github.com/ory/x@v0.0.513/prometheusx/middleware.go:39
github.com/urfave/negroni.middleware.ServeHTTP
        <MY_PATH>go/pkg/mod/github.com/urfave/negroni@v1.0.0/negroni.go:38 status:Bad Request status_code:400] http_request=map[headers:map[<COMMON_HEADER> cookie:[<CSRF_TOKEN_KEY><CSRF_TOKEN_VAL>=; ory_kratos_continuity=<COOKIE_TOKEN1>] dnt:1 sec-ch-ua:"<UA>] host:localhost:4433 method:GET path:/self-service/methods/oidc/callback/gitlab query:code=<MY_CODE>&state=<STATE_TOKEN1> remote:[::1]:60700 scheme:http] service_name=Ory Kratos service_version=master
INFO[2022-12-05T17:22:02+08:00]<MY_PATH>go/pkg/mod/github.com/ory/x@v0.0.513/reqlog/middleware.go:146 github.com/ory/x/reqlog.(*Middleware).ServeHTTP() completed handling request                    http_request=map[headers:map[<COMMON_HEADER> cookie:[<CSRF_TOKEN_KEY><CSRF_TOKEN_VAL>=; ory_kratos_continuity=<COOKIE_TOKEN1>] dnt:1 sec-ch-ua:"<UA>] host:localhost:4433 method:GET path:/self-service/methods/oidc/callback/gitlab query:code=<MY_CODE>&state=<STATE_TOKEN1> remote:[::1]:60700 scheme:http] http_response=map[headers:map[<CACHE_HEADER> location:http://localhost:3000/identity/error?id=f67cbd39-ab7c-4182-8166-1735d5be318d set-cookie:[ory_kratos_continuity=<TOKEN-t3rs>; Path=/; Expires=Wed, 04 Jan 2023 09:22:02 GMT; Max-Age=2592000; HttpOnly; SameSite=Lax] vary:Cookie] size:103 status:303 text_status:See Other took:7.164167ms]
INFO[2022-12-05T17:22:05+08:00]<MY_PATH>go/pkg/mod/github.com/ory/x@v0.0.513/reqlog/middleware.go:134 github.com/ory/x/reqlog.(*Middleware).ServeHTTP() started handling request                      http_request=map[headers:map[<COMMON_HEADER> cookie:[<CSRF_TOKEN_KEY><CSRF_TOKEN_VAL>=; ory_kratos_continuity=<TOKEN-t3rs>] dnt:1 origin:http://localhost:3000 referer:http://localhost:3000/ sec-ch-ua:"<UA>] host:localhost:4433 method:GET path:/self-service/errors query:id=f67cbd39-ab7c-4182-8166-1735d5be318d remote:[::1]:60700 scheme:http]
INFO[2022-12-05T17:22:05+08:00]<MY_PATH>go/pkg/mod/github.com/ory/x@v0.0.513/reqlog/middleware.go:146 github.com/ory/x/reqlog.(*Middleware).ServeHTTP() completed handling request                    http_request=map[headers:map[<COMMON_HEADER> cookie:[<CSRF_TOKEN_KEY><CSRF_TOKEN_VAL>=; ory_kratos_continuity=<TOKEN-t3rs>] dnt:1 origin:http://localhost:3000 referer:http://localhost:3000/ sec-ch-ua:"<UA>] host:localhost:4433 method:GET path:/self-service/errors query:id=f67cbd39-ab7c-4182-8166-1735d5be318d remote:[::1]:60700 scheme:http] http_response=map[headers:map[access-control-allow-credentials:true access-control-allow-origin:http://localhost:3000 access-control-expose-headers:Content-Type cache-control:private, no-cache, no-store, must-revalidate content-type:application/json; charset=utf-8 vary:Cookie] size:4964 status:200 text_status:OK took:13.471042ms]
```

some notable lines are 

1. first request to callback handler
```
INFO[2022-12-05T17:21:41+08:00]<MY_PATH>go/pkg/mod/github.com/ory/x@v0.0.513/reqlog/middleware.go:134 github.com/ory/x/reqlog.(*Middleware).ServeHTTP() started handling request                      http_request=map[headers:map[<COMMON_HEADER> cookie:[<CSRF_TOKEN_KEY><CSRF_TOKEN_VAL>=; ory_kratos_continuity=<COOKIE_TOKEN1>] dnt:1 sec-ch-ua:"<UA>] host:localhost:4433 method:GET path:/self-service/methods/oidc/callback/gitlab query:code=<MY_CODE>&state=<STATE_TOKEN1> remote:[::1]:60592 scheme:http]
```

2. completed handling request  
```
INFO[2022-12-05T17:22:02+08:00]<MY_PATH>go/pkg/mod/github.com/ory/x@v0.0.513/reqlog/middleware.go:146 github.com/ory/x/reqlog.(*Middleware).ServeHTTP() completed handling request                    http_request=map[headers:map[<COMMON_HEADER> cookie:[<CSRF_TOKEN_KEY><CSRF_TOKEN_VAL>=; ory_kratos_continuity=<COOKIE_TOKEN1>] dnt:1 sec-ch-ua:"<UA>] host:localhost:4433 method:GET path:/self-service/methods/oidc/callback/gitlab query:code=<MY_CODE>&state=<STATE_TOKEN1> remote:[::1]:60592 scheme:http] http_response=map[headers:map[<CACHE_HEADER> location:http://localhost:3000/ set-cookie:[ory_kratos_continuity=MTY3MDIzMjEwMXxEdi1CQkFFQ180SUFBUkFCRUFBQUJQLUNBQUE9fI1rWkD8bjafhmYvXPvhw2BkfK3pB5eC0hFLybCtbfLc; Path=/; Expires=Wed, 04 Jan 2023 09:21:41 GMT; Max-Age=2592000; HttpOnly; SameSite=Lax <CSRF_TOKEN_KEY>=LAAd83iH5n2fAJcEVEjGUmkZOy9HR0xQmo1L5HGiTKM=; Path=/; Max-Age=31536000; HttpOnly; SameSite=Lax ory_kratos_session=<SESSION_COOKIE>; Path=/; Expires=Tue, 06 Dec 2022 09:22:01 GMT; Max-Age=86399; HttpOnly; SameSite=Lax] vary:Cookie] size:49 status:303 text_status:See Other took:21.510034625s]
```

3. another request is handled again immediately after the first one was completed.

```
INFO[2022-12-05T17:22:02+08:00]<MY_PATH>go/pkg/mod/github.com/ory/x@v0.0.513/reqlog/middleware.go:134 github.com/ory/x/reqlog.(*Middleware).ServeHTTP() started handling request                      http_request=map[headers:map[<COMMON_HEADER> cookie:[<CSRF_TOKEN_KEY><CSRF_TOKEN_VAL>=; ory_kratos_continuity=<COOKIE_TOKEN1>] dnt:1 sec-ch-ua:"<UA>] host:localhost:4433 method:GET path:/self-service/methods/oidc/callback/gitlab query:code=<MY_CODE>&state=<STATE_TOKEN1> remote:[::1]:60700 scheme:http]
ERRO[2022-12-05T17:22:02+08:00]<MY_PATH>go/pkg/mod/github.com/ory/x@v0.0.513/logrusx/helper.go:125 github.com/ory/x/logrusx.(*Logger).Logf() An error occurred and is being forwarded to the error user interface.  audience=application error=map[debug:Resumable ID from cookie could not be found in the datastore: id=
```

Here are the corresponding Wireshark capture and browser console output.

![image](https://user-images.githubusercontent.com/21005146/205600911-5d8c1846-0008-49de-bd0d-9b7fcff713be.png)
![image](https://user-images.githubusercontent.com/21005146/205601067-2a83d39a-fa7f-4c20-943a-a446a49f0553.png)

### Version

0.11.0

### On which operating system are you observing this issue?

macOS

### In which environment are you deploying?

Binary

**Comment by CNLHC** [comment]
The method `/self-service/methods/oidc/callback/:provier` should not take too much time to respond because of: 

1. The UX is bad: the user clicks the `authorize` button on the provider's page and nothing happens for a while. The whole page is frozen to the user during the callback method exchanging token background.
2. The problem described in this issue may cause weird problems.

**Comment by CNLHC** [comment]
I make more tests on this problem.

using a simple stub handler in `strategy.go`

```
if handle, _, _ := r.Lookup("GET", "/self-service/methods/oidc/stub/:time"); handle == nil {
r.GET("/self-service/methods/oidc/stub/:time", func(w http.ResponseWriter, r *http.Request, p httprouter.Params) {
	stall := p.ByName("time")
	stall_second, err := strconv.Atoi(stall)
	if err != nil {
		w.WriteHeader(400)
		w.Write([]byte("invalid number format"))
		return
	}
	time.Sleep(time.Duration(stall_second) * time.Second)
	fmt.Println("sleep done, write 200 and hello")
	w.WriteHeader(200)
	_, err = w.Write([]byte("hello"))
	fmt.Println("write done, err: ", err)
})
}
```

1. `curl http://localhost:4433/self-service/methods/oidc/stub/9` will return normally
2. `curl http://localhost:4433/self-service/methods/oidc/stub/10` will return nothing (`curl: (52) Empty reply from server`)

After changing this line https://github.com/ory/kratos/blob/8406eaf92006d9812108bd3ae57245f01e627bfc/cmd/daemon/serve.go#L123
from `server := graceful.WithDefaults(&http.server{})` to ` server := (&http.Server{}) `  the `curl http://localhost:4433/self-service/methods/oidc/stub/10` will be ok.

**Comment by CNLHC** [comment]
So it seems a hardcoded `WriteTimeout` makes all the requests longer than 10 seconds impossible.

https://github.com/ory/graceful/blob/927f8152c87148e97688d02bc0abd44d75f4954a/http_defaults.go#L72

**Comment by aeneasr** [comment]
We have not seen this issue on production for a while, closing this as probably fixed.
